src/helpers/fee.ts

```diff
@@ -207,3 +207,143 @@ export function getSwapAmount(
+        MAX_FEE_NUMERATOR
+      )} bps`
+    );
+  }
+
+  if (targetFeeBps > cliffFeeBps) {
+    throw new Error(
+      "Target fee must be less than or equal to cliff fee for reduction"
+    );
+  }
+
+  const cliffFeeNumerator = bpsToFeeNumerator(cliffFeeBps);
+
+  const targetFeeNumerator = bpsToFeeNumerator(targetFeeBps);
+
+  const totalReduction = cliffFeeNumerator - targetFeeNumerator;
+  const reductionFactor = totalReduction / totalPeriods;
+
+  return Math.floor(reductionFactor);
+}
+
+/**
+ * Calculates the reduction factor for an Exponential fee schedule
+ * Formula: fee_numerator = cliff_fee_numerator * (1 - reduction_factor/BASIS_POINT_MAX)^passed_period
+ * reduction_factor = BASIS_POINT_MAX * (1 - (fee_numerator/cliff_fee_numerator)^(1/totalPeriods))
+ * @param cliffFeeBps - The initial fee value at the cliff in basis points
+ * @param targetFeeBps - The desired final fee value after all periods in basis points
+ * @param totalPeriods - The total number of periods over which the reduction occurs
+ * @returns The estimate of reduction factor
+ */
+export function estimateExponentialReductionFactor(
+  cliffFeeBps: number,
+  targetFeeBps: number,
+  totalPeriods: number
+): number {
+  // Validate inputs
+  if (totalPeriods <= 0) {
+    throw new Error("Total periods must be greater than zero");
+  }
+
+  if (cliffFeeBps > feeNumeratorToBps(MAX_FEE_NUMERATOR)) {
+    throw new Error(
+      `Cliff fee (${cliffFeeBps} bps) exceeds maximum allowed value of ${feeNumeratorToBps(
+        MAX_FEE_NUMERATOR
+      )} bps`
+    );
+  }
+
+  if (targetFeeBps > cliffFeeBps) {
+    throw new Error(
+      "Target fee bps must be less than or equal to cliff fee bps for reduction"
+    );
+  }
+
+  const cliffFeeNumerator = bpsToFeeNumerator(cliffFeeBps);
+
+  const targetFeeNumerator = bpsToFeeNumerator(targetFeeBps);
+
+  const ratio = targetFeeNumerator / cliffFeeNumerator;
+  const decayBase = Math.pow(ratio, 1 / totalPeriods);
+  const reductionFactor = BASIS_POINT_MAX * (1 - decayBase);
+  return Math.floor(reductionFactor);
+}
+
+/**
+ * Calculates the dynamic variable fee control parameter based on base fee, max volatility accumulator
+ *
+ * @param {BN} baseFeeBps - The base fee value in bps
+ * @param {BN} maxPriceChangeBps - The max price change between bin step in bps. Default 15%
+ * @param {number} binStep
+ * @returns {BN} The calculated dynamic variable fee control value
+ */
+export function getDynamicFeeParams(
+  baseFeeBps: number,
+  maxPriceChangeBps: number = 1500, // default 15%
+  binStep: number = 1
+): {
+  maxVolatilityAccumulator: BN;
+  variableFeeControl: BN;
+} {
+  const baseFeeNumerator = new BN(bpsToFeeNumerator(baseFeeBps));
+  const maxDynamicFeeNumerator = baseFeeNumerator.muln(20).divn(100); // default max dynamic fee = 15% of base fee.
+
+  const maxVolatilityAccumulator = new BN(BASIS_POINT_MAX * maxPriceChangeBps);
```

Comment:
we should validate if maxPriceChangeBps > 1500

Reply:
add validation